diff --git a/Android.bp b/Android.bp
index b08769d..230c659 100644
--- a/Android.bp
+++ b/Android.bp
@@ -29,9 +29,10 @@ rust_library {
     cargo_pkg_version: "0.2.0",
     srcs: ["src/lib.rs"],
     edition: "2018",
+    no_stdlibs: true,
     rustlibs: [
         "libbitflags",
-        "liblog_rust",
-        "libzerocopy",
+        "liblog_rust_nostd",
+        "libzerocopy_nostd",
     ],
     apex_available: [
