diff --git a/Android.bp b/Android.bp
index 2fffd32fb..eb06e7878 100644
--- a/Android.bp
+++ b/Android.bp
@@ -56,9 +56,9 @@ rust_library {
         "libspin",
         "libuntrusted",
     ],
-    static_libs: [
-        "libring_core_0_17_0_alpha_11_",
-        "libring_core_0_17_0_alpha_11_test",
+    whole_static_libs: [
+        "libring-core",
+        "libring-test",
     ],
     apex_available: [
         "//apex_available:platform",
@@ -94,9 +94,9 @@ rust_test {
         "libspin",
         "libuntrusted",
     ],
-    static_libs: [
-        "libring_core_0_17_0_alpha_11_",
-        "libring_core_0_17_0_alpha_11_test",
+    whole_static_libs: [
+        "libring-core",
+        "libring-test",
     ],
 }
 
