diff --git a/Cargo.toml b/Cargo.toml
index 13205cb..20147e1 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -63,7 +63,7 @@ version = "3"
 
 [features]
 alloc = [
-    "base64ct/alloc",
+    "base64ct?/alloc",
     "der/alloc",
 ]
 fingerprint = ["sha2"]
diff --git a/src/lib.rs b/src/lib.rs
index f466756..995f3bd 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -48,6 +48,10 @@ pub use crate::{
 };
 pub use der::{self, asn1::ObjectIdentifier};
 
+/// Local Android change: Use std to allow building as a dylib.
+#[cfg(android_dylib)]
+extern crate std;
+
 #[cfg(feature = "alloc")]
 pub use {crate::traits::EncodePublicKey, der::Document};
diff --git a/Cargo.toml.orig b/Cargo.toml.orig
index e7138a7..be1a9f6 100644
--- a/Cargo.toml.orig
+++ b/Cargo.toml.orig
@@ -26,7 +26,7 @@ hex-literal = "0.3"
 tempfile = "3"
 
 [features]
-alloc = ["base64ct/alloc", "der/alloc"]
+alloc = ["base64ct?/alloc", "der/alloc"]
 fingerprint = ["sha2"]
 pem = ["alloc", "der/pem"]
 std = ["der/std", "alloc"]
